### ISSUE TITLE
What is really loaded during one epoch? Can we deduce how many time the whole dataset is used?

### ISSUE THREAD

**Issue by naayem** [MIC-DKFZ/nnUNet#1292]
Hello everyone, I have an important question for a paper concerning the epochs. 
I am using the Totalsegmentator dataset and I am asked to define my epochs as the classical definition in deep learning. So from the epochs that I used I'm trying to infer how many times did I use the whole dataset.

The classical definition is as following:

> "An epoch is when all the training data is used at once and is defined as the total number of iterations of all the training data in one cycle for training the machine learning model."

 
But in the nature paper I read that there is a set of rules that defines an epoch. And there I get a little bit confused. 
Here are some elements from the paper:

> Distilling this knowledge into successful method design results in the following heuristic rule: “initialize the patch size to median image shape and iteratively reduce it while adapting the network topology accordingly (including network depth, number and position of pooling operations along each axis, feature map sizes and convolutional kernel sizes) until the network can be trained with a batch size of at least two given GPU memory constraints.” 

> To enable large patch sizes, the batch size of the networks in nnU-Net is small. In fact, most 3D U-Net confgurations were trained with a batch size of only two (Fig. SN5.1a in Supplementary Note 5). 

> Training schedule. Based on experience and as a trade-off between runtime and reward, all networks are trained for 1,000 epochs, with one epoch being defined as iteration over 250 mini-batches.

> Samples for the mini-batches are chosen from random training cases. Oversampling is implemented to ensure robust handling of class imbalances; 66.7% of samples are from random locations within the selected training case, while 33.3% of patches are guaranteed to contain one of the foreground classes that are present in the selected training sample (randomly selected). The number of foreground patches is rounded with a forced minimum of 1 (resulting in one random and one foreground patch with a batch size of two). A variety of data augmentation techniques are applied on the fly during training: rotations, scaling, Gaussian noise, Gaussian blur, brightness, contrast, simulation of low resolution, gamma correction and mirroring. Details are provided in Supplementary Note 4.

Okay, so for example I have a batch size of 2 and I run 4000 epochs on a training dataset of 1082 images. Normally there would be 541 iterations for one epoch, one epoch covering the whole dataset. But in the nature paper of 2020, one epoch is defined as iteration over 250 mini-batches. But mini-batches are chosen from random training cases. And there I'm lost in the explanation of mini-batches. Is the sampling totally random over the whole dataset or within the batch which is of size 2? Are we sure to cover the whole dataset?

In the 1018 paper one epoch is defined as :

> We define an epoch as the iteration over 250 training batches

I really did my best to answer this question alone. I also went through the code but I failed to understand of the sampling is done and I lacked of time despite spending an afternoon on it. If I a find an answer I will post it here.

Best regards and many thanks again for sharing the code.
Vincent

**Comment by FabianIsensee** [comment]
Hey,
the classic definition of epoch does not really make sense for patch-based segmentation. Think about it. If we are talking CT images, then your average image is maybe 512x512x512 voxels in size. Now you train with 128x128x128 patches, so when have you actually seen this training case? 
Implementing this the 'proper' way would be way too complex (sliding window over every image but still draw patches randomly and ideally switch the offset each epoch). That is just a pain and for what, really. There is no good reason to do it this way.
Yes, in principle, nnU-Net could adapt the number of epochs or the number of iterations per epoch to the dataset size. I have experimented with that and not found a god solution. What we are doing here works for 99% of datasets, so it's god enough I'd say.
Maybe it is easier to completely forego the concept of an epoch and just say we train for 1000*250 = 250k iterations.

Sampling is done the following way:
- pick a random case
- if oversampling:
  - pick a random class that is in that case
  - pick a random location where that class is
  - center the patch around that lcoation
- else:
  - pick a random location
  
We never check if all areas within all cases are sampled. We just assume that's the case
Best,
Fabian
